library/core/src/option.rs

```diff
@@ -735,22 +735,34 @@ impl<T> Option<T> {
         }
     }
 
-    const fn get_some_offset() -> isize {
-        if mem::size_of::<Option<T>>() == mem::size_of::<T>() {
-            // niche optimization means the `T` is always stored at the same position as the Option.
-            0
-        } else {
-            assert!(mem::size_of::<Option<T>>() == mem::size_of::<Option<mem::MaybeUninit<T>>>());
-            let some_uninit = Some(mem::MaybeUninit::<T>::uninit());
-            // SAFETY: This gets the byte offset of the `Some(_)` value following the fact that
-            // niche optimization is not active, and thus Option<T> and Option<MaybeUninit<t>> share
-            // the same layout.
-            unsafe {
-                (some_uninit.as_ref().unwrap() as *const mem::MaybeUninit<T>)
-                    .byte_offset_from(&some_uninit as *const Option<mem::MaybeUninit<T>>)
-            }
-        }
-    }
+    /// This is a guess at how many bytes into the option the payload can be found.
+    ///
+    /// For niche-optimized types it's correct because it's pigeon-holed to only
+    /// one possible place.  For other types, it's usually correct today, but
+    /// tweaks to the layout algorithm (particularly expansions of
+    /// `-Z randomize-layout`) might make it incorrect at any point.
+    ///
+    /// It's guaranteed to be a multiple of alignment (so will always give a
+    /// correctly-aligned location) and to be within the allocated object, so
+    /// is valid to use with `offset` and to use for a zero-sized read.
+    const SOME_BYTE_OFFSET_GUESS: isize = if mem::size_of::<Self>() == mem::size_of::<T>() {
+        // Zero is trivially a multiple of alignment and inside the object.
+        0
+    } else {
+        let some_uninit = Some(mem::MaybeUninit::<T>::uninit());
+        let offset =
+            // SAFETY: `as_ref` gives an address inside the existing `Option`,
+            // so both pointers are derived from the same thing and the result
+            // cannot overflow an `isize`.
+            unsafe { <*const _>::byte_offset_from(some_uninit.as_ref().unwrap(), &some_uninit) };
+
+        // These are all true by construction, but to make it obvious:
+        assert!(offset >= 0);
+        assert!((offset as usize) <= mem::size_of::<Self>());
```

Comment:
Yeah, same point as 8472 above, but I don't really see why this has to hold in the face of very silly layouts for `Option<MaybeUninit<T>>`. The other two I agree with.

Reply:
Yup, agreed with both of you.  Enough spurious padding and it could be out of bounds.

@rustbot author